src/celeritas/grid/SplineXsCalculator.hh

```diff
@@ -0,0 +1,265 @@
+
+        // number of grid indexs away from the specified energy that need to be
+        // checked in both directions
+        size_type order_steps = order_ / 2 + 1;
+
+        // true bounding indices of the grid that will be checked
+        size_type true_low_idx;
+        size_type true_high_idx;
+
+        // If the interpolation requests out-of-bounds indices, clip the
+        // extents. This will reduce the order of the interpolation
+        // todo - instead of clipping the bounds, alter both the low and high
+        // index to keep the range just shifted down
+        if (lower_idx >= order_steps - 1)
+        {
+            true_low_idx = lower_idx - order_steps + 1;
+        }
+        else
+        {
+            true_low_idx = 0;
+        }
+        if (lower_idx + order_steps < loge_grid_.size())
+        {
+            true_high_idx = lower_idx + order_steps;
+        }
+        else
+        {
+            true_high_idx = loge_grid_.size() - 1;
+        }
+
+        // if the requested interpolation order is even, a direction must be
+        // selected to interpolate to
+        if (order_ % 2 == 0)
+        {
+            real_type low_dist = loge - loge_grid_[lower_idx];
+            real_type high_dist = loge_grid_[lower_idx + 1] - loge;
+            CELER_ASSERT(low_dist >= 0);
+            CELER_ASSERT(high_dist >= 0);
+            if (low_dist > high_dist)
+            {
+                true_low_idx += 1;
+            }
+            else
+            {
+                true_high_idx -= 1;
+            }
+        }
+        result = this->interpolate(energy.value(), true_low_idx, true_high_idx);
+    }
+
+    return result;
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Get the cross section at the given index.
+ */
+CELER_FUNCTION real_type SplineXsCalculator::operator[](size_type index) const
+{
+    real_type energy = std::exp(loge_grid_[index]);
+    real_type result = this->get(index);
+
+    if (index >= data_.prime_index)
+    {
+        result /= energy;
+    }
+    return result;
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Get the raw cross section data at a particular index.
+ */
+CELER_FUNCTION real_type SplineXsCalculator::get(size_type index) const
+{
+    CELER_EXPECT(index < data_.value.size());
+    return reals_[data_.value[index]];
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Get the raw cross section data at a particular index.
+ */
+CELER_FUNCTION real_type SplineXsCalculator::interpolate(
+    real_type energy, size_type low_idx, size_type high_idx) const
+{
+    CELER_EXPECT(high_idx < loge_grid_.size());
+    real_type result = 0.0;
+
+    // Outer loop over indices for contributing to the result
+    for (size_type outer_idx = low_idx; outer_idx < high_idx + 1; ++outer_idx)
+    {
+        real_type outer_e = std::exp(loge_grid_[outer_idx]);
```

Comment:
Do you mean performing the interpolation in the log(E) space entirely or am I misunderstanding? Maybe an example on some actual xs functions/data would be helpful to me but I think this would just lead to an inaccurate interpolation. In testing it locally I found doing so on a linear function or even quadratic or exponential functions gives less accurate values than the version in this MR

Reply:
You're right, after #1450 I see that we should be interpolating in `E` to be consistent. We definitely *should* generate some cross sections...